File: src/Web/JavaScriptJS/patterns/interpreter.js
const assert = require("node:assert/strict");
/** Evaluates a tiny expression language through composable expression nodes. */
function run() {
  const number = (value) => ({ interpret: () => value });
  const add = (left, right) => ({ interpret: () => left.interpret() + right.interpret() });
  const multiply = (left, right) => ({ interpret: () => left.interpret() * right.interpret() });
  const expression = add(number(2), multiply(number(3), number(4)));
  assert.equal(expression.interpret(), 14);
  return "interpreter";
}
module.exports = { run };
if (require.main === module) run();

File: src/Other/MicroPython/interpreter.py
class Number:
    def __init__(self, value):
        self.value = value

    def interpret(self):
        return self.value


class Add:
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def interpret(self):
        return self.left.interpret() + self.right.interpret()


expression = Add(Add(Number(2), Number(3)), Number(4))
value = expression.interpret()
assert value == 9
print("interpreter={}".format(value))
